File: coding_theory/q10_locating_dominating_196/verify_q10_ld196.py
#!/usr/bin/env python3
"""Direct certificate checker for the 196-word LD code in Q_10."""

from __future__ import annotations

import collections
import itertools
import pathlib


DIMENSION = 10
VERTEX_COUNT = 1 << DIMENSION
CODE_PATH = pathlib.Path(__file__).with_name("code.txt")


def closed_neighborhood(vertex: int) -> frozenset[int]:
    return frozenset(
        {vertex, *(vertex ^ (1 << coordinate) for coordinate in range(DIMENSION))}
    )


def main() -> None:
    words = CODE_PATH.read_text(encoding="ascii").split()
    assert all(len(word) == DIMENSION and set(word) <= {"0", "1"} for word in words)
    code = frozenset(int(word, 2) for word in words)
    assert len(words) == len(code) == 196

    signatures: dict[frozenset[int], int] = {}
    signature_sizes: collections.Counter[int] = collections.Counter()
    incidence_count = 0
    for vertex in range(VERTEX_COUNT):
        if vertex in code:
            continue
        signature = code & closed_neighborhood(vertex)
        assert signature, f"undominated non-codeword {vertex:010b}"
        assert signature not in signatures, (
            f"duplicate signature for {signatures[signature]:010b} and {vertex:010b}"
        )
        signatures[signature] = vertex
        signature_sizes[len(signature)] += 1
        incidence_count += len(signature)

    assert len(signatures) == 828
    assert signature_sizes == {1: 162, 2: 403, 3: 201, 4: 55, 5: 7}

    distance_counts: collections.Counter[int] = collections.Counter(
        (first ^ second).bit_count()
        for first, second in itertools.combinations(sorted(code), 2)
    )
    expected_distances = {
        1: 67,
        2: 722,
        3: 2574,
        4: 3895,
        5: 4450,
        6: 4092,
        7: 2311,
        8: 781,
        9: 198,
        10: 20,
    }
    assert distance_counts == expected_distances
    assert sum(distance_counts.values()) == 196 * 195 // 2

    internal_edges = distance_counts[1]
    assert incidence_count == 1826
    assert incidence_count + len(code) + 2 * internal_edges == 11 * len(code)

    # Junnila--Laihonen--Lehtila: gamma^LD(Q_n) >= 2^(n+1)/(n+2), n >= 10.
    lower_bound = ((1 << (DIMENSION + 1)) + DIMENSION + 1) // (DIMENSION + 2)
    assert lower_bound == 171

    print("verified: 196-word locating-dominating code in Q_10")
    print("non-codewords: 828; distinct nonempty signatures: 828")
    print(
        "signature sizes:",
        " ".join(f"{size}:{signature_sizes[size]}" for size in sorted(signature_sizes)),
    )
    print(
        "pair distances:",
        " ".join(f"{distance}:{distance_counts[distance]}" for distance in range(1, 11)),
    )
    print("incidence check: 1826 + 196 + 2*67 = 11*196 = 2156")
    print("certified interval: 171 <= gamma^LD(Q_10) <= 196")


if __name__ == "__main__":
    main()

File: coding_theory/q10_locating_dominating_196/verify_q10_ld196.cpp
#include <array>
#include <bit>
#include <cassert>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <vector>

namespace {

constexpr int kDimension = 10;
constexpr int kVertexCount = 1 << kDimension;

int parse_word(const std::string& word) {
    assert(word.size() == kDimension);
    int value = 0;
    for (char bit : word) {
        assert(bit == '0' || bit == '1');
        value = 2 * value + (bit - '0');
    }
    return value;
}

int distance(int first, int second) {
    return std::popcount(static_cast<unsigned>(first ^ second));
}

}  // namespace

int main(int argc, char** argv) {
    const std::string path = argc > 1 ? argv[1] : "code.txt";
    std::ifstream input(path);
    assert(input);

    std::vector<int> code;
    std::array<bool, kVertexCount> is_code{};
    std::string word;
    while (input >> word) {
        const int value = parse_word(word);
        assert(!is_code[value]);
        is_code[value] = true;
        code.push_back(value);
    }
    assert(code.size() == 196);

    // This implementation deliberately uses direct Hamming-distance tests;
    // the Python checker generates XOR neighborhoods and intersects sets.
    std::set<std::vector<int>> signatures;
    std::array<int, kDimension + 2> signature_counts{};
    std::int64_t incidence_count = 0;
    for (int vertex = 0; vertex < kVertexCount; ++vertex) {
        if (is_code[vertex]) continue;
        std::vector<int> signature;
        for (int codeword : code) {
            if (distance(vertex, codeword) <= 1) signature.push_back(codeword);
        }
        assert(!signature.empty());
        assert(signatures.insert(signature).second);
        ++signature_counts[signature.size()];
        incidence_count += static_cast<std::int64_t>(signature.size());
    }
    assert(signatures.size() == 828);
    const std::array<int, 6> expected_signature_counts{0, 162, 403, 201, 55, 7};
    for (int size = 1; size <= 5; ++size) {
        assert(signature_counts[size] == expected_signature_counts[size]);
    }
    for (int size = 6; size < static_cast<int>(signature_counts.size()); ++size) {
        assert(signature_counts[size] == 0);
    }

    std::array<int, kDimension + 1> distance_counts{};
    for (std::size_t first = 0; first < code.size(); ++first) {
        for (std::size_t second = first + 1; second < code.size(); ++second) {
            ++distance_counts[distance(code[first], code[second])];
        }
    }
    const std::array<int, kDimension + 1> expected_distance_counts{
        0, 67, 722, 2574, 3895, 4450, 4092, 2311, 781, 198, 20,
    };
    assert(distance_counts == expected_distance_counts);
    int pair_count = 0;
    for (int count : distance_counts) pair_count += count;
    assert(pair_count == 196 * 195 / 2);

    assert(incidence_count == 1826);
    assert(incidence_count + 196 + 2 * distance_counts[1] == 11 * 196);

    constexpr int lower_bound = ((1 << 11) + 11) / 12;
    static_assert(lower_bound == 171);

    std::cout << "verified: 196-word locating-dominating code in Q_10\n"
              << "non-codewords: 828; distinct nonempty signatures: 828\n"
              << "signature sizes:";
    for (int size = 1; size <= 5; ++size) {
        std::cout << ' ' << size << ':' << signature_counts[size];
    }
    std::cout << "\npair distances:";
    for (int value = 1; value <= kDimension; ++value) {
        std::cout << ' ' << value << ':' << distance_counts[value];
    }
    std::cout << "\nincidence check: 1826 + 196 + 2*67 = 11*196 = 2156\n"
              << "certified interval: 171 <= gamma^LD(Q_10) <= 196\n";
}
